### PR TITLE
fix `push_down_filter` for pushing filters on grouping columns rather than aggregate columns

### LINKED ISSUE #4401: Optimization rule `filter_push_down` causes `FieldNotFound` error
**Describe the bug**
Optimization rule `filter_push_down` causes `FieldNotFound` error

**To Reproduce**
```sql
CREATE TABLE t AS VALUES (TIMESTAMP '2022-11-28T18:00:00');
SELECT date_trunc('day', column1) AS day FROM t GROUP BY day HAVING day IS NOT NULL;
SELECT date_trunc('day', column1) AS day FROM t GROUP BY day HAVING date_trunc('day', column1) IS NOT NULL;
```
Error displayed
```
SchemaError(FieldNotFound { field: Column { relation: None, name: "datetrunc(Utf8(\"day\"),t.column1)" }, valid_fields: Some([Column { relation: Some("t"), name: "column1" }]) })
```

**Expected behavior**
When I delete the optimization rule `filter_push_down`, the statement executes normally.
```
+---------------------+
| day                 |
+---------------------+
| 2022-11-28T00:00:00 |
+---------------------+
```
Here is the log about filter_push_down. I added some descriptions about the `Expr` type.

```
[2022-11-28T12:37:49Z DEBUG datafusion_optimizer::optimizer] eliminate_outer_join:
    Projection: Column(datetrunc(Utf8("day"),t.column1)) AS day
      Filter: Column(datetrunc(Utf8("day"),t.column1)) IS NOT NULL
        Aggregate: groupBy=[[ScalarFunction(datetrunc(Utf8("day"), Column(t.column1)))]], aggr=[[]]
          TableScan: t
    
[2022-11-28T12:37:49Z DEBUG datafusion_optimizer::optimizer] filter_push_down:
    Projection: Column(datetrunc(Utf8("day"),t.column1)) AS day
      Aggregate: groupBy=[[ScalarFunction(datetrunc(Utf8("day"), Column(t.column1)))]], aggr=[[]]
        Filter: Column(datetrunc(Utf8("day"),t.column1)) IS NOT NULL
          TableScan: t
    
```

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #4401.

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/optimizer/src/push_down_filter.rs [hunk 1]
```diff
@@ -643,18 +643,25 @@ impl OptimizerRule for PushDownFilter {
                 for expr in predicates {
                     let columns = expr.to_columns()?;
                     if columns.is_empty()
-                        || !columns
-                            .intersection(&used_columns)
-                            .collect::<HashSet<_>>()
-                            .is_empty()
```
Comment (jackwener) [inline datafusion/optimizer/src/push_down_filter.rs:649]: The original performance was bad

Comment (mingmwang) [reply datafusion/optimizer/src/push_down_filter.rs:649]: As I mentioned in the last PR, I think we do not need to check the aggregate Exprs, but just check the group by Exprs. In some cases, the same column can exist in both aggregate Exprs and group by Exprs, for example `select count(distinct col_a),  col_a from table  group by col_a; ` .  If there is a Filter applied to col_a, the Filter can still be pushed down even it is referred  by the agg Exprs.

Comment (mingmwang) [reply datafusion/optimizer/src/push_down_filter.rs:649]: The logic should check all the columns used by the Filter predicate is the subset of the group by Exprs output Columns.

Comment (jackwener) [reply datafusion/optimizer/src/push_down_filter.rs:649]: Yes. For push_down_filter through Agg, we can push `Expr` in `groupby_expr`.
Has add it.

File: datafusion/optimizer/src/push_down_filter.rs [hunk 2]
```diff
@@ -910,11 +922,9 @@ mod tests {
         // rewrite to CNF
         // (c = 1 OR c = 1) [can pushDown] AND (c = 1 OR b > 3) AND (b > 2 OR C = 1) AND (b > 2 OR b > 3)
 
-        let expected = "\
-        Filter: (test.c = Int64(1) OR b > Int64(3)) AND (b > Int64(2) OR test.c = Int64(1)) AND (b > Int64(2) OR b > Int64(3))\
+        let expected = "Filter: (test.c = Int64(1) OR test.c = Int64(1)) AND (test.c = Int64(1) OR b > Int64(3)) AND (b > Int64(2) OR test.c = Int64(1)) AND (b > Int64(2) OR b > Int64(3))\
```
Comment (liukun4515) [inline datafusion/optimizer/src/push_down_filter.rs:925]: cc @Ted-Jiang

File: datafusion/optimizer/src/push_down_filter.rs [hunk 3]
```diff
@@ -910,11 +922,9 @@ mod tests {
...
         \n  Aggregate: groupBy=[[test.a]], aggr=[[SUM(test.b) AS b]]\
-        \n    Filter: test.c = Int64(1) OR test.c = Int64(1)\
-        \n      TableScan: test";
```
Comment (jackwener) [inline datafusion/optimizer/src/push_down_filter.rs:917]: Original plan is wrong.😂
I think we need to delete this wrong UT.

Comment (Ted-Jiang) [reply datafusion/optimizer/src/push_down_filter.rs:917]: Yes, it was my fault.
`col_c` should not exist in filter.  Need delete it 😂
Is this case show you fix:  filter with agg result should not push down? @jackwener

**Comment by mingmwang** [comment]
LGTM.

**Comment by jackwener** [comment]
@alamb PTAL @Dandandan

File: datafusion/optimizer/src/push_down_filter.rs [hunk 4]
```diff
@@ -641,20 +633,27 @@ impl OptimizerRule for PushDownFilter {
...
+                    let cols = expr.to_columns()?;
+                    if cols.iter().all(|c| group_expr_columns.contains(c)) {
                         push_predicates.push(expr);
+                    } else {
+                        keep_predicates.push(expr);
                     }
                 }
 
-                let child = match conjunction(push_predicates) {
+                // As for plan Filter: Column(a+b) > 0 -- Agg: groupby:[Column(a)+Column(b)]
```
Comment (alamb) [inline datafusion/optimizer/src/push_down_filter.rs:644]: Nice -- this is getting quite sophisticated.

File: datafusion/optimizer/src/push_down_filter.rs [hunk 5]
```diff
@@ -641,20 +633,27 @@ impl OptimizerRule for PushDownFilter {
...
+                // After push, we need to replace `a+b` with Column(a)+Column(b)
+                // So we need create a replace_map, add {`a+b` --> Expr(Column(a)+Column(b))}
+                let mut replace_map = HashMap::new();
+                for expr in &agg.group_expr {
+                    replace_map.insert(expr.display_name()?, expr.clone());
```
Comment (alamb) [inline datafusion/optimizer/src/push_down_filter.rs:649]: Double checked that `display_name` is the right one: https://docs.rs/datafusion/14.0.0/datafusion/prelude/enum.Expr.html#method.display_name 👍

File: datafusion/optimizer/src/push_down_filter.rs [hunk 6]
```diff
@@ -881,40 +880,30 @@ mod tests {
...
+            .filter(col("test.b + test.a").gt(lit(10i64)))?
             .build()?;
-        // filter of aggregate is after aggregation since they are non-commutative
-        let expected = "\
-            Filter: b > Int64(10)\
-            \n  Aggregate: groupBy=[[test.a]], aggr=[[SUM(test.b) AS b]]\
-            \n    TableScan: test";
+        let expected =
+            "Aggregate: groupBy=[[test.b + test.a]], aggr=[[SUM(test.a), test.b]]\
+        \n  Filter: test.b + test.a > Int64(10)\
```
Comment (alamb) [inline datafusion/optimizer/src/push_down_filter.rs:890]: 👍  very nice

File: datafusion/optimizer/src/push_down_filter.rs [hunk 7]
```diff
@@ -881,40 +880,30 @@ mod tests {
...
-        // (c =1 AND b > 2) OR (c = 1 AND b > 3)
-        // rewrite to CNF
-        // (c = 1 OR c = 1) [can pushDown] AND (c = 1 OR b > 3) AND (b > 2 OR C = 1) AND (b > 2 OR b > 3)
-
         let expected = "\
-        Filter: (test.c = Int64(1) OR b > Int64(3)) AND (b > Int64(2) OR test.c = Int64(1)) AND (b > Int64(2) OR b > Int64(3))\
-        \n  Aggregate: groupBy=[[test.a]], aggr=[[SUM(test.b) AS b]]\
-        \n    Filter: test.c = Int64(1) OR test.c = Int64(1)\
-        \n      TableScan: test";
+            Filter: b > Int64(10)\
```
Comment (alamb) [inline datafusion/optimizer/src/push_down_filter.rs:904]: I agree this new plan is correct

File: datafusion/optimizer/tests/integration-test.rs [hunk 8]
```diff
@@ -304,6 +304,18 @@ fn join_keys_in_subquery_alias_1() {
...
 }
 
+#[test]
+fn push_down_filter_groupby_expr_contains_alias() {
+    let sql = "SELECT * FROM (SELECT (col_int32 + col_uint32) AS c, count(*) FROM test GROUP BY 1) where c > 3";
+    let plan = test_sql(sql).unwrap();
+    let expected = "Projection: c, COUNT(UInt8(1))\
+    \n  Projection: test.col_int32 + test.col_uint32 AS c, COUNT(UInt8(1))\
+    \n    Aggregate: groupBy=[[test.col_int32 + CAST(test.col_uint32 AS Int32)]], aggr=[[COUNT(UInt8(1))]]\
+    \n      Filter: test.col_int32 + CAST(test.col_uint32 AS Int32) > Int32(3)\
```
Comment (alamb) [inline datafusion/optimizer/tests/integration-test.rs:314]: 👍

**Review by alamb — APPROVED** [review]
Looks great to me -- thank you @jackwener

**Pull request merged**